### PR TITLE
Added kotlinx.json JsonElement serialization support

### PR DESCRIPTION
**Pull request by rozza**
JAVA-5239

### CODE REVIEW

File: bson-kotlinx/build.gradle.kts [hunk 1]
```diff
@@ -38,19 +38,27 @@ description = "Bson Kotlinx Codecs"
 
 ext.set("pomName", "Bson Kotlinx")
 
+java {
+    registerFeature("jsonSupport") {
```
Comment (rozza) [inline bson-kotlinx/build.gradle.kts:42]: Required to make the dependency optional - see: JAVA-5540

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt [hunk 2]
```diff
@@ -45,34 +50,93 @@ import org.bson.types.ObjectId
  *
  * For custom serialization handlers
  */
-public sealed interface BsonDecoder {
+@ExperimentalSerializationApi
+internal sealed interface BsonDecoder : Decoder, CompositeDecoder {
```
Comment (rozza) [inline bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt:54]: As this is sealed there is no need for it to have been public

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonEncoder.kt [hunk 3]
```diff
@@ -37,31 +38,57 @@ import org.bson.types.ObjectId
  *
  * For custom serialization handlers
  */
-public sealed interface BsonEncoder {
+@ExperimentalSerializationApi
+internal sealed interface BsonEncoder : Encoder, CompositeEncoder {
```
Comment (rozza) [inline bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonEncoder.kt:42]: Same as the BsonDecoder interface - its sealed and had no need to be public.

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/JsonBsonDecoder.kt [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import kotlinx.serialization.json.buildJsonObject
+import kotlinx.serialization.modules.SerializersModule
+import org.bson.AbstractBsonReader
+import org.bson.BsonBinarySubType
+import org.bson.BsonType
+import org.bson.UuidRepresentation
+import org.bson.internal.UuidHelper
+
+@OptIn(ExperimentalSerializationApi::class)
+internal interface JsonBsonDecoder : BsonDecoder, JsonDecoder {
```
Comment (rozza) [inline bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/JsonBsonDecoder.kt:37]: Provides Json decoding support

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/JsonBsonEncoder.kt [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import kotlinx.serialization.json.JsonPrimitive
+import kotlinx.serialization.json.double
+import kotlinx.serialization.json.int
+import kotlinx.serialization.json.long
+import kotlinx.serialization.modules.SerializersModule
+import org.bson.BsonWriter
+import org.bson.types.Decimal128
+
+@OptIn(ExperimentalSerializationApi::class)
+internal class JsonBsonEncoder(
```
Comment (rozza) [inline bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/JsonBsonEncoder.kt:35]: Provides Json encoding support

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodec.kt [hunk 6]
```diff
@@ -172,13 +172,13 @@ private constructor(
     }
 
     override fun encode(writer: BsonWriter, value: T, encoderContext: EncoderContext) {
-        serializer.serialize(DefaultBsonEncoder(writer, serializersModule, bsonConfiguration), value)
+        serializer.serialize(BsonEncoder.createBsonEncoder(writer, serializersModule, bsonConfiguration), value)
```
Comment (rozza) [inline bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodec.kt:175]: Use the factory to check which encoder / decoder to load.

File: bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt [hunk 7]
```diff
@@ -730,6 +735,67 @@ class KotlinSerializerCodecTest {
...
+                    put("double", 4.2)
+                    put("boolean", true)
+                    put("string", "String")
+                })
+
+        assertRoundTrips(expected, dataClass)
+    }
+
+    @Test
+    fun testDataClassWithNestedJsonElements() {
```
Comment (rozza) [inline bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt:781]: Added a test to ensure complex nested types also support json elements.

File: gradle/publish.gradle [hunk 8]
```diff
@@ -100,6 +100,7 @@ configure(javaProjects) { project ->
                 artifact sourcesJar
                 artifact javadocJar
 
+                suppressAllPomMetadataWarnings()
```
Comment (rozza) [inline gradle/publish.gradle:103]: A side effect of using gradle features.  We dont need to publish the metadata in the pom at this time - so I've disabled the warnings.

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonEncoder.kt (hunk 3 above)
Comment (rozza) [reply bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonEncoder.kt:42]: I'll revert as its not strictly true - users could in theory use this aside from the Codecs.

File: bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt (hunk 2 above)
Comment (rozza) [reply bson-kotlinx/src/main/kotlin/org/bson/codecs/kotlinx/BsonDecoder.kt:54]: I'll revert as its not strictly true - users could in theory use this aside from the Codecs.

File: bson-kotlinx/build.gradle.kts [hunk 9]
```diff
@@ -38,19 +38,23 @@ description = "Bson Kotlinx Codecs"
 
 ext.set("pomName", "Bson Kotlinx")
 
+java { registerFeature("jsonSupport") { usingSourceSet(sourceSets["main"]) } }
```
Comment (vbabanin) [inline bson-kotlinx/build.gradle.kts:41]: Has this functionality been tested with enabling an optional capability in a sample project?

Comment (rozza) [reply bson-kotlinx/build.gradle.kts:41]: The pom produced has the correct scope:

```
    <dependency>
      <groupId>org.jetbrains.kotlinx</groupId>
      <artifactId>kotlinx-serialization-json</artifactId>
      <scope>compile</scope>
      <optional>true</optional>
    </dependency>
```

File: bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt
Comment (vbabanin) [inline bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt:1]: It appears that the current tests may not be covering the JsonBsonEncoder, as the coverage analysis indicates that the class has not been accessed. Are these tests intended to evaluate the JsonBsonEncoder? If not, could we consider adding tests to ensure this functionality is covered by tests?

File: gradle/publish.gradle (hunk 8 above)
Comment (rozza) [reply gradle/publish.gradle]: Done

File: bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt
Comment (rozza) [reply bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt:1]: Thanks @vbabanin turns out that wasn't really being used at all, rather it was left to some other implementation of the `JsonEncoder`.  I overrode `encodeSerializableValue` to handle `JsonElement`s and added extra tests for coverage.

I also tested to make sure the encoding configuration was honored.

File: bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt [hunk 10]
```diff
@@ -753,6 +771,219 @@ class KotlinSerializerCodecTest {
...
+                    put("documentEmpty", buildJsonObject {})
+                    put("document", buildJsonObject { put("a", JsonPrimitive(1)) })
+                    put("double", JsonPrimitive(62.0))
+                    put("int32", JsonPrimitive(42))
+                    put("int64", JsonPrimitive(52))
+                    put("objectId", JsonPrimitive("211111111111111111111112"))
+                    put("string", JsonPrimitive("the fox ..."))
+                    put("timestamp", JsonPrimitive(1311768464867721221))
+                })
+        assertDecodesTo("""{"value": $jsonAllSupportedTypesDocument}""", dataClassWithAllSupportedJsonTypes)
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson-kotlinx/src/test/kotlin/org/bson/codecs/kotlinx/KotlinSerializerCodecTest.kt:998]: Done - it is not roundtrippable so encodes and decodes can be different.

**Comment by rozza** [comment]
@vbabanin going to pause until #1462 is merged then rebase before final review as I think there will be some code share between the two.

**Pull request merged**